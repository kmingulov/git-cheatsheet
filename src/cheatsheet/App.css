.header {
  padding-left: 15px !important;
  padding-right: 15px !important;
}

.mainContent {
  padding-top: 20px;
}

.git-command-column {
  width: 40%;
}

/* .git { } */

/* .git-command { } */

/* .git-remote { } */

.git-url {
  text-decoration: underline;
}

.git-branch {
  font-weight: bold;
}

/* .git-commit { } */

/* .git-ref { } */

.git-file {
  font-style: italic;
}
