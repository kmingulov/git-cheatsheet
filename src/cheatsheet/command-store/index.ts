export * from './CommandStore';
export * from './lunr';
