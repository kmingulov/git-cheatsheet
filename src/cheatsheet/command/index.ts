export * from './Command';
export * from './CommandGroup';
export * from './CommandPart';
export * from './CommandPartType';
