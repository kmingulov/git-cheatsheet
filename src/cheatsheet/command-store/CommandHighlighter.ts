import { Command, CommandPart, CommandPartType } from 'cheatsheet/command';

const highlightPart = (part: CommandPart, regexp: RegExp): CommandPart[] =>
  part.text.split(regexp)
    .map(item =>
      item.match(regexp)
        ? new CommandPart(CommandPartType.HIGHLIGHT, item)
        : new CommandPart(part.role, item),
    );

const highlightParts = (parts: CommandPart[], regexp: RegExp): CommandPart[] =>
  parts
    .map(part => highlightPart(part, regexp))
    .reduce((theseParts, otherParts) => theseParts.concat(otherParts), []);

export class CommandHighlighter {

  public static highlight(command: Command, terms: string[]): Command {
    const regexp = new RegExp('(' + terms.join('|') + ')', 'i');

    const highlightedCommand = highlightParts(command.command, regexp);
    const highlightedDescription = highlightParts(command.description, regexp);
    const highlightedShortcut = highlightParts(command.scmBreezeShortcut, regexp);

    return new Command(command.id, highlightedCommand, highlightedDescription, highlightedShortcut);
  }

  public static highlightAll(commands: Command[], terms: string[]): Command[] {
    return commands.map(command => this.highlight(command, terms));
  }

}
