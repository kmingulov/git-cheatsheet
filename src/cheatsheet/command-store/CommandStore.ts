import { Command, CommandGroup } from 'cheatsheet/command';

export interface CommandStore {
  getAll(): CommandGroup[];
  search(searchQuery: string): Command[];
}
