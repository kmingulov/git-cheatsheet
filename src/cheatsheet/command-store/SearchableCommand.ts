import { Command } from '../command/Command';
import { CommandPart } from '../command/CommandPart';

const partsToString = (parts: CommandPart[], delimeter: string): string =>
  parts.map(part => part.text).join(delimeter);

export class SearchableCommand {
  public readonly id: string;
  public readonly command: string;
  public readonly description: string;
  public readonly scmBreezeShortcut: string;

  constructor(command: Command) {
    this.id = command.id;
    this.command = partsToString(command.command, ' ');
    this.description = partsToString(command.description, '');
    this.scmBreezeShortcut = partsToString(command.scmBreezeShortcut, ' ');
  }
}
