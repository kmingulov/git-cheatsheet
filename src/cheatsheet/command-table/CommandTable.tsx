import React, { Component, ReactElement } from 'react';

import { CommandGroup } from '../command/CommandGroup';
import { CommandGroupTable } from './CommandGroupTable';

interface CommandTableProps {
  groups: CommandGroup[];
}

class CommandTable extends Component<CommandTableProps> {
  public render(): ReactElement {
    return (
      <div>
        { this.props.groups.map(group => <CommandGroupTable group={ group } key={ group.title }/>) }
      </div>
    );
  }
}

export default CommandTable;
