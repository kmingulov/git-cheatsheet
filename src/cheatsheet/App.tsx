import React, { Component, ReactElement } from 'react';
import { Container, FormControl, InputGroup, Navbar } from 'react-bootstrap';
import { FaSearch } from 'react-icons/fa';

import { CommandStore } from './command-store';
import { CommandTable } from './command-table';

import commands from './command-data/commands';

import './App.css';

const store = new CommandStore(commands);

const initialState = { searchTerm: '' };
type State = Readonly<typeof initialState>;

class App extends Component<object, State> {
  public readonly state: State = initialState;

  public render(): ReactElement {
    return (
      <div>
        <Navbar bg='primary' sticky='top'>
          <Container className='header'>
            <Navbar.Brand>Git Cheatsheet</Navbar.Brand>
          </Container>
        </Navbar>

        <Container className='mainContent'>
          <InputGroup className='searchBox'>
            <InputGroup.Prepend>
              <InputGroup.Text>
                <FaSearch/>
              </InputGroup.Text>
            </InputGroup.Prepend>
              <FormControl
                placeholder='Searchâ€¦'
                onChange={ (event: any) => this.setState({ searchTerm: event.target.value }) }
              />
          </InputGroup>

          <CommandTable store={ store } searchTerm={ this.state.searchTerm }/>
        </Container>
      </div>
    );
  }
}

export default App;
