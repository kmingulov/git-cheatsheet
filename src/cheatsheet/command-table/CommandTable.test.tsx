import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';

import { Command, CommandGroup } from 'cheatsheet/command';
import { CommandStore } from 'cheatsheet/command-store';

import { CommandTable } from './CommandTable';

Enzyme.configure({adapter: new Adapter()});

jest.mock('./CommandGroupTable');

class CommandStoreStub implements CommandStore {
  public getAll(): ReadonlyArray<CommandGroup> {
    return [
      new CommandGroup('basic', 'Basic Commands', []),
      new CommandGroup('advanced', 'Advanced Commands', []),
    ];
  }
  public search(searchTerm: string): ReadonlyArray<Command> {
    return [ new Command('cmd-1', [], []) ];
  }
}

describe(CommandTable, () => {
  let store: CommandStore;
  let getAllSpy: jest.SpyInstance;
  let searchSpy: jest.SpyInstance;

  beforeEach(() => {
    store = new CommandStoreStub();
    getAllSpy = jest.spyOn(store, 'getAll');
    searchSpy = jest.spyOn(store, 'search');
  });

  it('renders full list of commands when searchTerm isn\'t present', () => {
    const component = mount(<CommandTable store={ store }/>);

    expect(component).toMatchSnapshot();
    expect(getAllSpy).toHaveBeenCalledTimes(1);
    expect(searchSpy).not.toHaveBeenCalled();
  });

  it('renders found commands when searchTerm is present', () => {
    const component = mount(<CommandTable store={ store } searchTerm='search'/>);

    expect(component).toMatchSnapshot();
    expect(getAllSpy).not.toHaveBeenCalled();
    expect(searchSpy).toHaveBeenCalledTimes(1);
    expect(searchSpy).toHaveBeenCalledWith('search');
  });
});
