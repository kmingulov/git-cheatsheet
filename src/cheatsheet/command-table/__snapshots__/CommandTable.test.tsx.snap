// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommandTable renders found commands when searchTerm is present 1`] = `
<CommandTable
  searchTerm="search"
  store={
    CommandStoreStub {
      "getAll": [MockFunction],
      "search": [MockFunction] {
        "calls": Array [
          Array [
            "search",
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": Array [
              Command {
                "command": Array [],
                "description": Array [],
                "id": "cmd-1",
                "scmBreezeShortcut": Array [],
              },
            ],
          },
        ],
      },
    }
  }
>
  <div>
    <CommandGroupTable
      group={
        CommandGroup {
          "commands": Array [
            Command {
              "command": Array [],
              "description": Array [],
              "id": "cmd-1",
              "scmBreezeShortcut": Array [],
            },
          ],
          "title": "Search Results",
        }
      }
      key="Search Results"
    />
  </div>
</CommandTable>
`;

exports[`CommandTable renders full list of commands when searchTerm isn't present 1`] = `
<CommandTable
  store={
    CommandStoreStub {
      "getAll": [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": Array [
              CommandGroup {
                "commands": Array [],
                "title": "Basic Commands",
              },
              CommandGroup {
                "commands": Array [],
                "title": "Advanced Commands",
              },
            ],
          },
        ],
      },
      "search": [MockFunction],
    }
  }
>
  <div>
    <CommandGroupTable
      group={
        CommandGroup {
          "commands": Array [],
          "title": "Basic Commands",
        }
      }
      key="Basic Commands"
    />
    <CommandGroupTable
      group={
        CommandGroup {
          "commands": Array [],
          "title": "Advanced Commands",
        }
      }
      key="Advanced Commands"
    />
  </div>
</CommandTable>
`;
